plugins/inputs/socketstat/socketstat.go

```
@@ -0,0 +1,188 @@
+	// Check that ss is installed
+	ssPath, err := exec.LookPath("ss")
+	if err != nil {
+		return new(bytes.Buffer), err
+	}
+
+	// Add needed args
+	cmdName := ssPath
+	var args []string
+	args = append(args, "-in")
+	args = append(args, "--"+proto)
+
+	// Run ss, return the output as bytes.Buffer
+	cmd := exec.Command(cmdName, args...)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	err = internal.RunTimeout(cmd, Timeout.Duration)
+	if err != nil {
+		return &out, fmt.Errorf("error running ss -in --%s: ", err)
+	}
+	return &out, nil
+}
+
+var validFields = "(bytes_acked|bytes_received|segs_out|segs_in|data_segs_in|data_segs_out)"
+var validValues = regexp.MustCompile("^" + validFields + ":[0-9]+$")
+var beginsWithBlank = regexp.MustCompile("^\\s+.*$")
+
+func (ss *Socketstat) parseAndGather(data *bytes.Buffer, proto string, acc telegraf.Accumulator) error {
+	scanner := bufio.NewScanner(data)
+	tags := map[string]string{}
+	fields := make(map[string]interface{})
+
+	// ss output can have blank lines, and/or socket basic info lines and more advanced
+	// statistics lines, in turns.
+	// We're using the flushData variable to determine if we should add a new measurement
+	// or postpone it to a later line
+
+	// The first line is only headers
+	scanner.Scan()
+
+	flushData := false
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "" {
+			continue
+		}
+		words := strings.Fields(line)
+
+		var err error
+		if !beginsWithBlank.MatchString(line) {
+			if flushData {
+				acc.AddFields(measurement, fields, tags)
+				flushData = false
+			}
+			// Delegate the real parsing to getTagsAndState, which manages various
+			// formats depending on the protocol
+			tags, fields = getTagsAndState(proto, words)
+			flushData = true
+		} else {
+			for _, word := range words {
+				if validValues.MatchString(word) {
+					kv := strings.Split(word, ":")
+					fields[kv[0]], err = strconv.ParseUint(kv[1], 10, 64)
+					if err != nil {
+						continue
```

You might want to output the error somewhere as otherwise we might silently drop fields.

---

Please mark as resolved.